include/ghex/glue/atlas/atlas_user_concepts.hpp

```diff
@@ -0,0 +1,467 @@
+                const atlas::Field& partition() const noexcept { return m_partition; }
+                const atlas::Field& remote_index() const noexcept { return m_remote_index; }
+                std::size_t levels() const noexcept { return m_levels; }
+                index_t size() const noexcept { return m_size; }
+                /** @brief first local index, excluding halo points*/
+                index_t first() const noexcept { return m_first; }
+                /** @brief last local index, excluding halo points*/
+                index_t last() const noexcept { return m_last; }
+
+                // print
+                /** @brief print */
+                template<class CharT, class Traits>
+                friend std::basic_ostream<CharT, Traits>& operator << (std::basic_ostream<CharT, Traits>& os, const atlas_domain_descriptor& domain) {
+                    os << "domain id = " << domain.domain_id() << ";\n"
+                       << "size = " << domain.size() << ";\n"
+                       << "# levels = " << domain.levels() << ";\n"
+                       << "partition indexes: [" << domain.partition() << "]\n"
+                       << "remote indexes: [" << domain.remote_index() << "]\n";
+                    return os;
+                }
+
+        };
+
+        /** @brief halo generator for atlas domains
+         * An Atlas domain has already the notion of halos.
+         * The purpose of the Atlas halo generator is to fill a container
+         * with indexes referring to remote partitions.
+         * @tparam DomainId domain id type*/
+        template<typename DomainId>
+        class atlas_halo_generator {
+
+            public:
+
+                // member types
+                using domain_type = atlas_domain_descriptor<DomainId>;
+                using index_t = atlas::idx_t;
+
+                /** @brief essentially a partition index and a sequence of remote indexes;
+                 * conceptually, it is similar to an iteration space
+                 * (only differences are that here it is already specialized for atlas
+                 * and provides an insert method as well),
+                 * and this is why the two concepts could potentially coincide.
+                 * WARN: size() gives only horizontal size.*/
+                class halo {
+
+                    private:
+
+                        int m_partition;
+                        std::vector<index_t, gridtools::ghex::allocator::unified_memory_allocator<index_t>> m_local_index;
+                        std::vector<index_t> m_remote_index;
+                        std::size_t m_levels;
+
+                    public:
+
+                        // ctors
+                        halo() noexcept = default;
+                        halo(const int partition, const std::size_t levels = 1) noexcept :
+                            m_partition{partition},
+                            m_local_index{},
+                            m_remote_index{},
```

Comment:
Yep, it's only because I'm used to list all the member initializations explicitly (here is trivial, but in general I find it clearer). I can remove it if you think it's too much.